kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaEvictionService.java

```diff
@@ -0,0 +1,164 @@
+ * replicas.
+ */
+public class ReplicaEvictionService extends AbstractScheduledService {
+  private static final Logger LOG = LoggerFactory.getLogger(ReplicaEvictionService.class);
+
+  private final CacheSlotMetadataStore cacheSlotMetadataStore;
+  private final ReplicaMetadataStore replicaMetadataStore;
+  private final KaldbConfigs.ManagerConfig managerConfig;
+  private final MeterRegistry meterRegistry;
+
+  @VisibleForTesting protected int futuresListTimeoutSecs = DEFAULT_ZK_TIMEOUT_SECS;
+
+  public static final String REPLICA_MARK_EVICT_SUCCEEDED = "replica_mark_evict_succeeded";
+  public static final String REPLICA_MARK_EVICT_FAILED = "replica_mark_evict_failed";
+  public static final String REPLICA_MARK_EVICT_TIMER = "replica_mark_evict_timer";
+
+  protected final Counter replicaMarkEvictSucceeded;
+  protected final Counter replicaMarkEvictFailed;
+  private final Timer replicaMarkEvictTimer;
+
+  public ReplicaEvictionService(
+      CacheSlotMetadataStore cacheSlotMetadataStore,
+      ReplicaMetadataStore replicaMetadataStore,
+      KaldbConfigs.ManagerConfig managerConfig,
+      MeterRegistry meterRegistry) {
+    this.cacheSlotMetadataStore = cacheSlotMetadataStore;
+    this.replicaMetadataStore = replicaMetadataStore;
+    this.managerConfig = managerConfig;
+    this.meterRegistry = meterRegistry;
+
+    // schedule configs checked as part of the AbstractScheduledService
+
+    replicaMarkEvictSucceeded = meterRegistry.counter(REPLICA_MARK_EVICT_SUCCEEDED);
+    replicaMarkEvictFailed = meterRegistry.counter(REPLICA_MARK_EVICT_FAILED);
+    replicaMarkEvictTimer = meterRegistry.timer(REPLICA_MARK_EVICT_TIMER);
+  }
+
+  @Override
+  protected void runOneIteration() {
+    markReplicasForEviction();
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(
+        managerConfig.getScheduleInitialDelayMins(),
+        managerConfig.getReplicaEvictionServiceConfig().getSchedulePeriodMins(),
+        TimeUnit.MINUTES);
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting replica eviction service");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Closed replica eviction service");
+  }
+
+  /**
+   * Marks cache slots that contain an expired replica as ready for eviction. Once eviction has
+   * started the cache node will further update the state indicating it has started evicting the
+   * replica from memory/disk.
+   */
+  protected int markReplicasForEviction() {
+    Timer.Sample evictionTimer = Timer.start(meterRegistry);
+
+    Map<String, ReplicaMetadata> replicaMetadataByReplicaId =
+        replicaMetadataStore
+            .getCached()
+            .stream()
+            .collect(Collectors.toUnmodifiableMap(ReplicaMetadata::getName, Function.identity()));
+
+    long expireAfterNow = Instant.now().toEpochMilli();
```

Comment:
If the `expireAfterNow` time is passed in as an argument from the caller function, it makes this function deterministic and easier to unit test.

Reply:
Done.